File: java/src/main/java/com/harness/Orchestrator.java
package com.harness;

import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public class Orchestrator {

    public List<LlmApi.LlmResponse> consensus(List<LlmRouter> routers, LlmApi.LlmRequest request) throws Exception {

        List<CompletableFuture<LlmApi.LlmResponse>> futures = routers.stream().map(router -> {
            LlmRouter r = new LlmRouter(router.getProvider(), router.getApiKey());
            return r.sendRequestAsync(request);
        }).collect(Collectors.toList());

        CompletableFuture<Void> allFutures = CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));

        try {
            allFutures.join();
        } catch (Exception e) {
            throw new Exception("One or more consensus requests failed", e);
        }

        return futures.stream()
                .map(CompletableFuture::join)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }

    public LlmApi.LlmResponse race(List<LlmRouter> routers, LlmApi.LlmRequest request) throws Exception {
        if (routers == null || routers.isEmpty()) {
            throw new IllegalArgumentException("No routers provided for race loop");
        }

        CompletableFuture<LlmApi.LlmResponse> firstSuccess = new CompletableFuture<>();
        AtomicInteger failures = new AtomicInteger(0);
        int totalRouters = routers.size();

        List<CompletableFuture<LlmApi.LlmResponse>> futures = routers.stream().map(router -> {
            LlmRouter r = new LlmRouter(router.getProvider(), router.getApiKey());
            CompletableFuture<LlmApi.LlmResponse> future = r.sendRequestAsync(request);

            future.whenComplete((result, ex) -> {
                if (ex != null) {
                    if (failures.incrementAndGet() == totalRouters) {
                        firstSuccess.completeExceptionally(new Exception("All race requests failed"));
                    }
                } else {
                    firstSuccess.complete(result);
                }
            });
            return future;
        }).collect(Collectors.toList());

        try {
            LlmApi.LlmResponse result = firstSuccess.join();

            for (CompletableFuture<LlmApi.LlmResponse> future : futures) {
                if (!future.isDone()) {
                    future.cancel(true);
                }
            }
            return result;
        } catch (Exception e) {
             throw new Exception("All race requests failed", e);
        }
    }
}

File: csharp/Harness/Orchestrator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Harness.LlmApi
{
    public class Orchestrator
    {
        public async Task<List<LlmResponse>> ConsensusAsync(List<LlmRouter> routers, LlmRequest request, CancellationToken cancellationToken = default)
        {
            var tasks = routers.Select(router =>
            {
                var safeRouter = new LlmRouter(router.Provider, router.ApiKey);
                return safeRouter.SendRequestAsync(request, cancellationToken);
            }).ToList();

            try
            {
                var results = await Task.WhenAll(tasks);
                return results.ToList();
            }
            catch (Exception)
            {
                var innerExceptions = tasks.Where(t => t.IsFaulted).Select(t => t.Exception).ToList();
                throw new AggregateException("One or more consensus requests failed", innerExceptions);
            }
        }

        public async Task<LlmResponse> RaceAsync(List<LlmRouter> routers, LlmRequest request, CancellationToken cancellationToken = default)
        {
            if (routers == null || !routers.Any())
            {
                throw new ArgumentException("No routers provided for race loop");
            }

            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);

            var tasks = routers.Select(async router =>
            {
                var safeRouter = new LlmRouter(router.Provider, router.ApiKey);
                var result = await safeRouter.SendRequestAsync(request, cts.Token);
                return result;
            }).ToList();

            var exceptions = new List<Exception>();

            while (tasks.Any())
            {
                var finishedTask = await Task.WhenAny(tasks);
                tasks.Remove(finishedTask);

                if (finishedTask.Status == TaskStatus.RanToCompletion)
                {
                    cts.Cancel();
                    return await finishedTask;
                }
                else if (finishedTask.IsFaulted)
                {
                    exceptions.Add(finishedTask.Exception);
                }
            }

            throw new AggregateException("All race requests failed", exceptions);
        }
    }
}
